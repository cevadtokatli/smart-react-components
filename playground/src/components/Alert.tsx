import { Alert, AlertIcon, AlertCloseIcon } from '@smart-react-components/ui'
import React from 'react'

export const Icon = () => (
  <>
    <Alert marginBottom="$length.3">
      <div>Lorem ipsum dolor sit amet</div>
      <AlertIcon hasHover><AlertCloseIcon /></AlertIcon>
    </Alert>
  </>
)

export const Paragraph = () => (
  <>
    <Alert marginBottom="$length.3">
      <h4>Title</h4>
      <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent tincidunt rhoncus nulla, vitae cursus ipsum egestas eget. In et pulvinar nibh. Integer scelerisque neque eu tellus luctus, non lobortis neque hendrerit. Proin commodo velit nec est gravida, in tempus risus faucibus. Aliquam sapien diam, rhoncus ut tempus sed, lobortis a nulla. Vestibulum pretium vehicula nisl in tempor. Mauris vel condimentum ante. Maecenas in erat ligula. Nunc in ullamcorper turpis, nec rhoncus turpis. Fusce ac auctor tellus. Nullam semper finibus arcu. Duis id odio rhoncus, suscipit sapien aliquam, tempor mi. In sit amet tortor finibus, pulvinar nibh ac, porttitor odio. Donec tincidunt sapien et sapien tristique auctor. Suspendisse lorem risus, efficitur id dui a, convallis vehicula risus.</p>
      <p>Quisque in augue vitae libero pellentesque dictum. Mauris facilisis, libero sed posuere ullamcorper, arcu nibh luctus neque, quis vulputate nisl lacus non ligula. Morbi a orci velit. Fusce lectus dolor, ullamcorper ut ante non, pulvinar commodo lectus. Phasellus viverra pharetra tortor, et vehicula est. Quisque in mi nec purus ultricies blandit. Etiam vel consequat metus, ac pellentesque felis. Morbi nec mi hendrerit erat dapibus egestas. Pellentesque id odio ac nunc commodo rhoncus at eu eros. Duis sodales, eros ac ultrices sodales, purus arcu dapibus nisl, id lobortis mauris massa in ligula. Nam eu augue in eros porttitor ornare sit amet eget lorem. Praesent porttitor tellus eu sodales posuere. Nulla dignissim, turpis at imperdiet posuere, mi massa congue diam, sit amet fringilla metus nulla eget dui.</p>
      <hr />
      <p>Phasellus pretium a felis eget aliquam. Vestibulum consequat diam dignissim sollicitudin fermentum. Suspendisse risus leo, suscipit tristique arcu ac, accumsan consequat risus. Phasellus elit sem, elementum id est nec, lacinia tempus augue. Aliquam nec odio porttitor, placerat tellus sed, fermentum odio. Interdum et malesuada fames ac ante ipsum primis in faucibus. Vestibulum vitae imperdiet odio. Nam venenatis hendrerit mollis. In hac habitasse platea dictumst. Vivamus dictum porttitor sem, in mattis ante euismod posuere. Vivamus et elementum est. Quisque porta varius arcu, eu dapibus lorem molestie eget. Vestibulum ut nulla blandit, facilisis turpis at, aliquet libero.</p>
      <p>Sed eget dui in ex sollicitudin placerat. Nulla feugiat sit amet massa sit amet ullamcorper. Ut tempus id felis et feugiat. Aliquam placerat purus eu orci lobortis elementum. Cras id tristique quam, vitae fringilla magna. Aenean eu metus rhoncus, elementum eros eu, porttitor tortor. In id vestibulum magna. Nunc iaculis eros eget mi efficitur commodo quis nec diam. Nullam felis nisi, cursus dapibus libero nec, iaculis tempor purus. Maecenas non sollicitudin urna, eget dignissim leo. Donec id nibh sodales turpis congue malesuada in at massa.</p>
      <p>Suspendisse congue rhoncus augue. <a>Curabitur</a> id metus mi. Curabitur condimentum dignissim justo ut suscipit. Ut molestie velit sed elit finibus, vel volutpat felis euismod. Donec tristique sagittis justo sed interdum. Nam sit amet molestie massa, ut vestibulum sapien. Mauris tincidunt vel nulla sed scelerisque. In nec tortor at mi luctus elementum non at diam. Duis nec imperdiet quam, in euismod erat. Donec venenatis tortor mauris, nec efficitur lectus commodo et. Nam accumsan nunc dolor, ac finibus risus ornare sed. Duis placerat ultrices est eu feugiat. Etiam purus metus, elementum volutpat cursus quis, ultrices at erat. Quisque varius semper nisi sit amet bibendum. Sed molestie vestibulum justo faucibus interdum.</p>  
    </Alert>
  </>
)

export const Palette = () => (
  <>
    <Alert marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="secondary" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="success" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="danger" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="warning" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="info" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="dynamic" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="!dynamic" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
  </>
)

export const Outline = () => (
  <>
    <Alert isOutline marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="secondary" isOutline marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="success" isOutline marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="danger" isOutline marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="warning" isOutline marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="info" isOutline marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="dynamic" isOutline marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="!dynamic" isOutline marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
  </>
)

export const Shape = () => (
  <>
    <Alert marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert shape="rounded" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
  </>
)

export const Size = () => (
  <>
    <Alert size="small" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert size="large" marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
  </>
)

export const Soft = () => (
  <>
    <Alert isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="secondary" isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="success" isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="danger" isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="warning" isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="info" isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="dynamic" isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="!dynamic" isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
  </>
)

export const SoftOutline = () => (
  <>
    <Alert isOutline isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="secondary" isOutline isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="success" isOutline isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="danger" isOutline isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="warning" isOutline isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="info" isOutline isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="dynamic" isOutline isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
    <Alert palette="!dynamic" isOutline isSoft marginBottom="$length.3">Lorem ipsum dolor sit amet</Alert>
  </>
)
