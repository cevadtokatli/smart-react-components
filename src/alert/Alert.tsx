import React from "react"
import styled from "styled-components"
import ElementProps from "../element/Props"
import {SizeStrictProps} from "../props/size-strict-props"
import Div from "../element/Div"
import {SizeProps} from "../props/size-props"
import {JSXElementProps, JSXChildren} from "../props"
import DV from "../default-value"
import IconElement from "./IconElement"

interface AlertElementProps extends ElementProps, SizeStrictProps {
    type?: string
    fill$: boolean
    icon: boolean
    iconType: string
    iconPosition: string
}

const getCustomKeys = fill => ({
    size: (v, t) => `
        padding: ${t.alert.padding[v].y} ${t.alert.padding[v].x};
        font-size: ${t.alert.fontSize[v]};

        > ${IconElement} {
            ${fill
            ? `
                padding: 0 ${t.alert.padding[v].xHalf};
                margin-top: -${t.alert.padding[v].y};
                margin-bottom: -${t.alert.padding[v].y};

                &[data-src-icon-position="left"] {
                    margin-right: ${t.alert.padding[v].xHalf};
                    margin-left: -${t.alert.padding[v].x};
                }
    
                &[data-src-icon-position="right"] {
                    margin-right: -${t.alert.padding[v].x};
                    margin-left: ${t.alert.padding[v].xHalf};
                }
            `
            : `
                &[data-src-icon-position="left"] {
                    margin-right: ${t.alert.padding[v].xHalf};
                    margin-left: -${t.alert.padding[v].xHalf};
                }

                &[data-src-icon-position="right"] {
                    margin-right: -${t.alert.padding[v].xHalf};
                    margin-left: ${t.alert.padding[v].xHalf};
                }
            `}
        }
    `
})

const AlertElement = styled(Div).attrs<AlertElementProps>(({size="default",type="primary",fill$,iconType,mb="3"}) => ({
    customKeys: getCustomKeys(fill$),
    size,
    type,
    iconType: iconType || type,
    mb
}))<AlertElementProps>(({theme,type,fill$,icon,iconType}) => `
    box-sizing: border-box;
    border: solid 1px;
    border-radius: ${theme.src.alert.radius};
    word-wrap: break-word;
    word-break: break-word;
    font-family: ${theme.src.fontFamily.alert};

    ${fill$
    ? `
        border-color: ${theme.src.type[type].main};
        background: ${theme.src.type[type].main};
        color: ${theme.src.type[type].alert.fontFill};    
        fill: ${theme.src.type[type].alert.fontFill};

        a {
            color: ${theme.src.type[type].alert.fontFill};    
        }

        hr {
            background: ${theme.src.type[type].darkest};
        }
    `
    : `
        border-left-width: 10px;
        border-color: ${theme.src.type[type].alert.border};
        background: ${theme.src.type[type].alert.background};
        color: ${theme.src.type[type].alert.font};   
        fill: ${theme.src.type[type].alert.font}; 

        a {
            color: ${theme.src.type[type].alert.linkFont};
        }

        hr {
            background: ${theme.src.type[type].alert.hr};
        }
    `}
    
    ${icon ? `
        display: flex;

        > ${IconElement} {
            ${fill$ ? `
                background: ${iconType == type ? theme.src.type[iconType].darker : theme.src.type[iconType].main};
                fill: ${theme.src.type[iconType].alert.fontFill};
            ` : `
                fill: ${iconType == type ? theme.src.type[iconType].alert.font : theme.src.type[iconType].main};
            `}
        }

        .src-alert-content {
            flex: 1 1 auto;
        }
    ` : ""}

    > {
        header, h1, h2, h3, h4, h5, h6, p, hr {
            margin-top: 0;
            margin-bottom: ${theme.src.length["3"]};

            &:last-child {
                margin-bottom: 0;
            }
        }
    }

    a {
        font-weight: ${theme.src.alert.linkFontWeight};

        &:hover {
            text-decoration: underline;
        }
    }
    
    hr {
        width: 100%;
        height: 1px;
        border: 0;
    }
`)

export interface Props extends SizeProps {
    elementProps?: JSXElementProps
    children: JSXChildren
    type?: string
    fill?: boolean
    icon?: JSX.Element
    iconType?: string
    iconPosition?: string
}

const Alert: React.FC<Props> = ({size="default",sizeSm,sizeMd,sizeLg,sizeXl,elementProps=DV.JSX_ELEMENT_PROPS,children,type,fill,icon,iconType,iconPosition="left"}) => {
    const IconEl = icon ? <IconElement data-src-icon-position={iconPosition}>{icon}</IconElement> : null

    return (
        <AlertElement {...elementProps} size={size} sizeSm={sizeSm} sizeMd={sizeMd} sizeLg={sizeLg} sizeXl={sizeXl} type={type} fill$={fill} icon={icon ? true : false} iconType={iconType} iconPosition={iconPosition}>
            {iconPosition == "left" && IconEl}
            {icon ? <div className="src-alert-content">{children}</div> : children}
            {iconPosition == "right" && IconEl}
        </AlertElement>
    )
}

export default Alert