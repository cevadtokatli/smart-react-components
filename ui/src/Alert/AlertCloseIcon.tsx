import CloseIcon from '../icons/Close'

export default CloseIcon
