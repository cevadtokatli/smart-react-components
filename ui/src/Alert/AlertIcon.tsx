import extractElementProps from '@smart-react-components/core/element-props'
import clickEvents, { ClickEvents } from '@smart-react-components/core/element-props/click-events'
import { PaletteProp, ResponsiveProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import AlertIconElement from '../components/Alert/AlertIconElement'
import { OrderPosition } from '../types'

export interface Props extends ClickEvents {
  children: JSX.Element
  hasBackground?: boolean
  hasHover?: boolean
  palette?: PaletteProp
}

interface PrivateProps extends ResponsiveProp<'size', SizeProp> {
  alertPalette: PaletteProp
  isOutline: boolean
  isSoft: boolean
  position: OrderPosition
}

const AlertIcon: React.FC<Props> = (props: Props & PrivateProps) => (
  <AlertIconElement
    alertPalette={props.alertPalette ?? props.palette}
    alertIconSize={props.size}
    alertIconSizeSm={props.sizeSm}
    alertIconSizeMd={props.sizeMd}
    alertIconSizeLg={props.sizeLg}
    alertIconSizeXl={props.sizeXl}
    hasBackground={props.hasBackground}
    hasHover={props.hasHover}
    iconPosition={props.position}
    isOutline={props.isOutline}
    isSoft={props.isSoft}
    palette={props.palette ?? props.alertPalette}
    {...extractElementProps(props, [clickEvents])}
  >{props.children}</AlertIconElement>
)

AlertIcon.defaultProps = {
  hasBackground: true,
}

AlertIcon.displayName = 'SRCAlertIcon'

export default AlertIcon
