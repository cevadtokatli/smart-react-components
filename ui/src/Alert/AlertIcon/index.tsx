import extractElementProps from '@smart-react-components/core/element-props'
import clickEvents, { ClickEvents } from '@smart-react-components/core/element-props/click-events'
import { PaletteProp, ResponsiveProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import AlertIconElement from './AlertIconElement'

export interface Props extends ClickEvents {
  children: JSX.Element
  hasBackground?: boolean
  hasHover?: boolean
  palette?: PaletteProp
}

interface PrivateProps extends ResponsiveProp<'size', SizeProp> {
  alertPalette: PaletteProp
  isSoft: boolean
}

const AlertIcon: React.FC<Props> = (props: Props & PrivateProps) => (
  <AlertIconElement
    alertPalette={props.alertPalette}
    alertIconSize={props.size}
    alertIconSizeSm={props.sizeSm}
    alertIconSizeMd={props.sizeMd}
    alertIconSizeLg={props.sizeLg}
    alertIconSizeXl={props.sizeXl}
    hasBackground={props.hasBackground}
    hasHover={props.hasHover}
    isSoft={props.isSoft}
    palette={props.palette ?? props.alertPalette}
    {...extractElementProps(props, [clickEvents])}
  >{props.children}</AlertIconElement>
)

AlertIcon.displayName = 'SRCIcon'

export default AlertIcon
