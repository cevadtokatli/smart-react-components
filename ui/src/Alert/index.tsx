import extractElementProps from '@smart-react-components/core/element-props'
import clickEvents, { ClickEvents } from '@smart-react-components/core/element-props/click-events'
import intrinsicStyledProps, { IntrinsicStyledProps } from '@smart-react-components/core/element-props/intrinsic-styled-props'
import useChangeEffect from '@smart-react-components/core/hooks/useChangeEffect'
import { ContentElement, JSXElementProps, PaletteProp, Partial, ResponsiveProp, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import AlertElement, { Content } from './AlertElement'

export interface Props extends
  Partial<ResponsiveProp<'size', SizeProp>>,
  ClickEvents,
  IntrinsicStyledProps {
  children: ContentElement | ContentElement[]
  elementProps?: JSXElementProps
  isOutline?: boolean
  isSoft?: boolean
  palette?: PaletteProp
  shape?: ShapeProp
}

const Alert: React.FC<Props> = props => {
  const getContent = () => {
    const children = !Array.isArray(props.children) ? [props.children] : props.children
    const content = []
    let iconLeft = null
    let iconRight = null

    for (let i = 0; i < children.length; i++) {
      const item = children[i] as JSX.Element
      if (item.type?.displayName === 'SRCAlertIcon') {
        const iconEl = React.cloneElement(item, {
          alertPalette: props.palette,
          isSoft: props.isSoft,
          size: props.size,
          sizeSm: props.sizeSm,
          sizeMd: props.sizeMd,
          sizeLg: props.sizeLg,
          sizeXl: props.sizeXl,
        })

        if (i === 0) {
          iconLeft = iconEl
        } else if (i === children.length - 1) {
          iconRight = iconEl
        }

        continue
      }

      content.push(typeof children[i] === 'string' ? children[i] : React.cloneElement(item, { key: item.key ?? i }))
    }

    return {
      children: (
        <>
          { iconLeft && iconLeft }
          <Content>{content}</Content>
          { iconRight && iconRight }
        </>
      ),
      hasIconLeft: !!iconLeft,
      hasIconRight: !!iconRight,
    }
  }

  const [{ children, hasIconLeft, hasIconRight }, setContent] = React.useState(() => getContent())

  useChangeEffect(() => {
    setContent(getContent())
  }, [props.children])

  return (
    <AlertElement
      alertSize={props.size}
      alertSizeSm={props.sizeSm}
      alertSizeMd={props.sizeMd}
      alertSizeLg={props.sizeLg}
      alertSizeXl={props.sizeXl}
      hasIconLeft={hasIconLeft}
      hasIconRight={hasIconRight}
      isOutline={props.isOutline}
      isSoft={props.isSoft}
      palette={props.palette}
      shape={props.shape}
      {...extractElementProps(props, [clickEvents, intrinsicStyledProps])}
    >
      {children}
    </AlertElement>
  )
}

Alert.defaultProps = {
  palette: 'primary',
  shape: 'rectangle',
  size: 'medium',
}

export default Alert
