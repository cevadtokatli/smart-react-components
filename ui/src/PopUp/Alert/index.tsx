import useChangeEffect from '@smart-react-components/core/hooks/useChangeEffect'
import { Theme } from '@smart-react-components/core/theme'
import { ContentElement, PaletteProp, SetState } from '@smart-react-components/core/types'
import { TransitionAfterCallback, TransitionBeforeCallback } from '@smart-react-components/transition/types'
import React from 'react'
import { ThemeContext } from 'styled-components'
import Button from '../../Button'
import Modal from '../../Modal'
import ModalBody from '../../Modal/ModalBody'
import ModalFooter from '../../Modal/ModalFooter'
import ModalHeader from '../../Modal/ModalHeader'
import ModalTitle from '../../Modal/ModalTitle'

export interface Props {
  action?: () => (void | Promise<void>)
  afterHide?: TransitionAfterCallback
  afterShow?: TransitionAfterCallback
  beforeHide?: TransitionBeforeCallback
  beforeShow?: TransitionBeforeCallback
  buttonLabel?: string
  buttonPalette?: PaletteProp
  hasHideAnimation?: boolean
  hasShowAnimation?: boolean
  hasOverlayBackground?: boolean
  hasOverlayBlurEffect?: boolean
  message: ContentElement
  setStatus?: SetState<boolean>
  status?: boolean
  transitionClassName?: string
  transitionDuration?: number
  title?: ContentElement
}

const Alert: React.FC<Props> = ({ action, afterHide, afterShow, beforeHide, beforeShow, buttonLabel, buttonPalette, hasHideAnimation, hasShowAnimation, hasOverlayBackground, hasOverlayBlurEffect, message, setStatus, status, transitionClassName, transitionDuration, title }) => {
  const theme = React.useContext<Theme>(ThemeContext)
  const [isClicked, setClicked] = React.useState(false)

  useChangeEffect(() => {
    if (status) {
      setClicked(false)
    }
  }, [status])

  const handleClick = async () => {
    if (isClicked) {
      return
    }

    setClicked(true)

    const result = action?.()

    if (result instanceof Promise) {
      await result
    }

    setStatus?.(false)
  }

  return (
    <Modal
      afterHide={afterHide}
      afterShow={afterShow}
      beforeHide={beforeHide}
      beforeShow={beforeShow}
      hasHideAnimation={hasHideAnimation}
      hasShowAnimation={hasShowAnimation}
      hasOverlayBackground={hasOverlayBackground}
      hasOverlayBlurEffect={hasOverlayBlurEffect}
      isDismissible={false}
      setStatus={setStatus}
      status={status}
      transitionClassName={transitionClassName}
      transitionDuration={transitionDuration}
    >
      { title && (
        <ModalHeader>
          <ModalTitle>{title}</ModalTitle>
        </ModalHeader>
      ) }
      <ModalBody>{message}</ModalBody>
      <ModalFooter hasBorder={false} display="flex" justifyContent="flex-end">
        <Button
          onClick={handleClick}
          palette={buttonPalette}
        >
          {buttonLabel ?? theme.$.i18n.ok}
        </Button>
      </ModalFooter>
    </Modal>
  )
}

export default Alert
