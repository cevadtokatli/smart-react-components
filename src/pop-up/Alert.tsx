import React from "react"
import PopUpProps from "./PopUpProps"
import {JSXChild} from "../props"
import {AlertLanguage} from "../types/pop-up"
import DEFAULT_LANGUAGE from "../constants/pop-up/alert-language"
import DOMHelper from "../helper/DOMHelper"
import PopUp from "./PopUp"
import ModalHeader from "../modal/Header"
import ModalTitle from "../modal/Title"
import ModalBody from "../modal/Body"
import ModalFooter from "../modal/Footer"
import Button from "../button/Button"

export interface Props extends PopUpProps {
    type?: string
    title?: JSXChild
    message: JSXChild
    callback?: () => void
    lang?: AlertLanguage
}

const Alert: React.FC<Props> = ({modalElementProps,overlayProps,type="primary",title,message,callback,lang=DEFAULT_LANGUAGE,status,setStatus,transitionClassName,transitionDuration,transitionType,showAnimation,hideAnimation,beforeOpen,beforeClose,afterOpen,afterClose}) => {
    React.useEffect(() => {
        if(status) {
            DOMHelper.addEventListener(document.body, ["keydown"], keydown)

            return () => {
                DOMHelper.removeEventListener(document.body, ["keydown"], keydown)
            }
        }
    }, [status])

    const close = () => {
        if(callback)
            callback()

        if(setStatus)
            setStatus(false)
    }

    const keydown = (e:KeyboardEvent) => {
        switch(e.keyCode) {
            case 13:
            case 27:
                close()
        }
    }
    
    return (
        <PopUp modalElementProps={modalElementProps} overlayProps={overlayProps} status={status} transitionClassName={transitionClassName} transitionDuration={transitionDuration} transitionType={transitionType} showAnimation={showAnimation} hideAnimation={hideAnimation} beforeOpen={beforeOpen} beforeClose={beforeClose} afterOpen={afterOpen} afterClose={afterClose}>
            {title && <ModalHeader><ModalTitle>{title}</ModalTitle></ModalHeader>}
            <ModalBody>{message}</ModalBody>
            <ModalFooter pt={[0, true]}>
                <Button type={type} link={DOMHelper.isMobile} waveEffect={DOMHelper.isMobile ? "dark" : "light"} elementProps={{onClick:close}}>{lang.close}</Button>
            </ModalFooter>
        </PopUp>
    )
}

export default Alert