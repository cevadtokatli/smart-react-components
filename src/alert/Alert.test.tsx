import React from "react"
import renderTheme from "../test/renderTheme"
import mountTheme from "../test/mountTheme"
import Alert from "./Alert"

describe("<Alert />", () => {
    it("should match snapshot", () => {
        const wrapper = renderTheme(<Alert>Alert</Alert>)
        expect(wrapper).toMatchSnapshot()
    })

    describe("icon", () => {
        it("should include src-alert-content class", () => {
            const wrapper = mountTheme(<Alert icon={<i />}>Alert</Alert>)
            expect(wrapper.find(".src-alert-content").exists()).toBe(true)
        })
    
        it("should render icon", () => {
            const wrapper = mountTheme(<Alert icon={<i id="icon" />}>Alert</Alert>)
            expect(wrapper.find("#icon").exists()).toBe(true)
        })
    
        it("should render icon at the right of the child element", () => {
            const wrapper = mountTheme(<Alert icon={<i id="icon" />} iconPosition="right">Alert</Alert>)
            expect(wrapper.find(".src-alert-content + * #icon").exists()).toBe(true)
        })
    })
})