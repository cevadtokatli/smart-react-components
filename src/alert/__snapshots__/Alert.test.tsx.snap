// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> should match snapshot 1`] = `
<div
  class="sc-AxjAm sc-AxiKw hELdDk"
  type="primary"
>
  Alert
</div>
`;
